Adding support to `DiracDelta` for generic points in the domain

kishore-nori opened the pull request:
Hi @santiagobadia and @amartinhuertas, I added methods to `DiracDelta` to support generic points in the domain, which are not grid entities with tags, would be glad to know your suggestions for improvement and changes. This PR is aimed at resolving the issue [#357]. Given that `CellField`s can now be evaluated at a generic point in the domain, the method make use of this to calculate the integral of the composition of `DiracDelta` with a `CellField`, and also works for generic functions compatible to evaluate at a `Gridap` `Point`. 

So as to conform with `DiracDelta` struct, a `DiscreteModel` is constructed with given set of `Point`s to generate `Triangulation` and `Measure`. If we are not required to conform with the existing framework, we could just store the `Point`s as a struct and evaluate the `CellField` point-wise directly, but the above doesn't create much overhead. And probably the later is not generic enough to be extendable to line and surface forcing terms. 

The usage is same as earlier but the constructor is as follows:

```julia
model = CartesianDiscreteModel((0,1,0,1),(3,3))
p = Point(0.1,0.2)
pvec = [p, 1.5*p]
δ = DiracDelta(model,p)
δ_comb = DiracDelta(model,pvec)
```

[WIP] TO DO: support for line and surface loads! (like presented here https://github.com/gridap/Gridap.jl/issues/408#issuecomment-734929575). I think the line load case is extendable in a straight forward manner using `CartesianDiscreteModel` (but may be needs to be done separately for each line - defined by a pair of points..)

amartinhuertas commented:
HI @kishore-nori ! Thanks for your work. Can you please add an entry to the `NEWS.md` file pointing to this work ?

amartinhuertas commented on test/CellDataTests/DiracDeltasTests.jl:
```diff
@@ -37,6 +37,22 @@ degree = 2
 
+# Tests for DiracDelta at a generic Point in the domain #
+
+p = VectorValue(0.2,0.3)
```
why do you use `VectorValue` and not `Point`?

kishore-nori replied:
`Point` is not available for usage here, is has been defined in `Gridap.Fields` so just used `VectorValue` (as it seemed to me that the tests are also following the module hierarchy).

amartinhuertas replied:
I dont see any problem in `using Gridap.Fields` at the test. (perhaps I am missing something)

kishore-nori replied:
No problem will use it now, I was in a misconception that even the tests follow the same module include hierarchy as the Gridap source code does.

amartinhuertas commented on test/CellDataTests/DiracDeltasTests.jl:
```diff
@@ -37,6 +37,22 @@ degree = 2
+
+p = VectorValue(0.2,0.3)
+δ = DiracDelta(p,model)
+@test sum(δ(v)) == v(p)
```
Replace `==` by `\sim` ?

kishore-nori replied:
Yes I can do that, should be safer..

amartinhuertas commented on test/CellDataTests/DiracDeltasTests.jl:
```diff
@@ -37,6 +37,22 @@ degree = 2
+
+pvec = [p,π*p]
+δ = DiracDelta(pvec,model)
+@test sum(δ(v)) == sum(v(pvec))
```
see https://github.com/gridap/Gridap.jl/pull/837/files#r994035461

kishore-nori replied:
(replied above)

amartinhuertas commented on test/TensorValuesTests/IndexingTests.jl:
```diff
@@ -14,6 +14,7 @@ v = VectorValue{4}(a)
 @test size(v) == (4,)
 @test length(v) == 4
 @test lastindex(v) == length(v)
+@test v[end] == a[end]
```
where do these tests come from? shouldnt be these already in `master` ? (i.e., is your branch up to date with `master`?) they seem to be unrelated to this PR.

kishore-nori replied:
No, I forgot adding these tests, sorry took the opportunity to add it now (instead of again creating a PR), sorry for the confusion, it is not under PR.

amartinhuertas replied:
ok, no worries ... we may merge them here, but the cleaner way to go is to do it in a separate PR

kishore-nori replied:
Yes I am sorry, will be careful next time and separate it into different PR. Probably would be too messy here now, to revert back?

amartinhuertas replied:
Better not to revert back, I agree. 

For completeness, perhaps we can add in the NEWS.md a comment that we added some missing tests missing from another PR.

kishore-nori replied:
Definitely, will mention it in the `NEWS.md`

amartinhuertas commented on test/CellDataTests/DiracDeltasTests.jl:
```diff
@@ -37,6 +37,22 @@ degree = 2
+δ = DiracDelta(pvec,model)
+@test sum(δ(v)) == sum(v(pvec))
+
+# below passes but cannot use FESpace and FEFunction here
```
Not sure why this commented code is here. Should it be here or we can just remove it?

kishore-nori replied:
This is a good test but as I mentioned in the comment, I cant use `FESpace` and `FEFunction` here, as they are down the hierarchy. If it is ok I can use `using Gridap` for the test

amartinhuertas replied:
>  I cant use FESpace and FEFunction here, as they are down the hierarchy. 

I dont understand this comment. Can't you just add `using Gridap.FESpaces` to the module in that test file?

kishore-nori replied:
Yes I can.., my question was does `Gridap` `tests` also follow the module usage hierarchy as the `src`? If not I can do it.

amartinhuertas replied:
I dont think we have to constrain ourselves to the same module usage hierarchy. In Gridap this has to be respected because otherwise we may end with dependcy loops (cycles) among modules.

You can think as the test directory as an external program that is using Gridap tools and features. Thus, I do not see why we are forced to respect that constraint.

amartinhuertas replied:
The other thing that we have to think is whether `DiracDeltas.jl` should still be within `CellData` now given that we have introduced additional dependencies.

kishore-nori replied:
> The other thing that we have to think is whether DiracDeltas.jl should still be within CellData now given that we have introduced additional dependencies.
No.. I haven't introduced any dependency in DiracDeltas.jl, used only the existing dependencies

amartinhuertas replied:
> No.. I haven't introduced any new dependency in DiracDeltas.jl, used only the existing dependencies

ok, I think you are right. Having a (new) reference to `CellField` (in the new `evaluate` function) does not imply a new module dependency, correct?

kishore-nori replied:
Yes.. since `CellField` belongs to `CellData` module itself, it is not a new or external dependency.

amartinhuertas commented on src/CellData/DiracDeltas.jl:
```diff
@@ -76,3 +79,33 @@ end
+  DiracDelta{0,NotGridEntity}(point_trian,dx)
+end
+
+function evaluate!(cache,d::DiracDelta{0,NotGridEntity},f::CellField)
```
why do we need this specialized version of `evaluate!`. Doesn't the generic one work straight away?

kishore-nori replied:
No.. it results in quadrature and `FEFunction/CellField` not having the same/compatible background triangulation error

amartinhuertas replied:
Ok. I now understand. The integrand is defined on a domain which is not (currently) compatible with the domain of the measure. But this may mean that there is a compatibility rule among triangulations that is somehow missing, because why these domains should not be compatible?

amartinhuertas replied:
This is just a though ... it you have the impression that this overly complicates thing, we may stick into the current workaround.

kishore-nori replied:
Initially I tried to make it work with the same triangulations but it wasn't generalising properly due to `CellQuadrature` constructing extra-cells of `#undef` values in `DomainContribution`. 

The problem occurs at @check is_change_possible(trian_f,trian_x) msg in the integrate method (https://github.com/gridap/Gridap.jl/blob/2a4921b473fe113b9d4b6168b8728a7a708f7bc6/src/CellData/CellQuadratures.jl#L142)

kishore-nori replied:
> But this may mean that there is a compatibility rule among triangulations that is somehow missing, because why these domains should not be compatible?
Yes I was wondering about this as well.. will have to dig deeper

amartinhuertas commented on test/CellDataTests/DiracDeltasTests.jl:
```diff
@@ -37,6 +37,22 @@ degree = 2
+# uh = FEFunction(V,rand(num_free_dofs(V)))
+# @test sum(d(uh)) == uh(p)
+
 #using Gridap
```
should we also test DiracDeltas with `FEFunction` and `FEBasis` ?

kishore-nori replied:
Yes, liked I replied to one of the above comments, we need to include the corresponding modules to use the `FEFunction` and `FEBasis`, I would be happy to include these tests..

amartinhuertas reviewed (changes requested):
Hi @kishore-nori ! I have finished a first round of review. Please address or rebut my comments/remarks/etc..

kishore-nori commented:
Thank you very much for the comments Alberto, I ll make changes accordingly following our discussion in comments here.

kishore-nori commented:
Hi @amartinhuertas and @santiagobadia, I have rewritten the implementation (realised I could do it in a good way, after our discussion to approach it as quadrature). This now works for evaluations and also solving the PDE with `DiracDelta` at a single or even a set of `Points`. I added tests to compare with both the exact solution and the earlier tag based implementation of `DiracDelta`. I also incorporated the earlier review comments from our discussions. Would be happy to know further comments for improvement.

amartinhuertas commented on src/CellData/DiracDeltas.jl:
```diff
@@ -65,14 +70,63 @@ function DiracDelta{0}(model::DiscreteModel;tags)
+  GenericDiracDelta{0,D,NotGridEntity}(trianv,pmeas)
+end
+
+function evaluate!(cache,d::GenericDiracDelta{0,Dt,NotGridEntity},f::Function) where Dt
```
Which is the ultimate reason behind having to define this specialised method? 

Why cant we leverage the generic implementation of `evaluate!` whenever the domain of `DiracDelta` is a function?

kishore-nori replied:
I was in a misconception that existing `evaluate!` method goes through the search process of Point to Cell (which ofcourse works perfectly fine as well), which is only necessary for `CellFields`, Functions can be evaluated directly without the search. But it seems it doesn't go through that route once we have the `CellPoint` available, which is very nice. So I ll remove it..

kishore-nori replied:
But one advantage of the new constructor is it is ~4 times faster and aloocates ~10 times less memory.

amartinhuertas replied:
> But it seems it doesn't go through that route once we have the CellPoint available, which is very nice.

Whenever you have a `CellPoint`, Gridap assumes that the points in each cell are within the cell (either reference cell or physical cells, depending on the value of the corresponding trait). 

If the points are not within the cell, then I guess you have an undefined behaviour. But we are satisfying this precondition, as the cells have been judiciously selected in the TriangulationView such that the hold the corresponding points.

amartinhuertas replied:
> But one advantage of the new constructor is it is ~4 times faster and aloocates ~10 times less memory.

Do these ratios keep asympotatically fixed as you refine the mesh?

kishore-nori replied:
> Whenever you have a CellPoint, Gridap assumes that the points in each cell are within the cell (either reference cell or physical cells, depending on the value of the corresponding trait).

>If the points are not within the cell, then I guess you have an undefined behaviour. But we are satisfying this precondition, as the cells have been judiciously selected in the TriangulationView such that they hold the corresponding points.

Yes true.. since we are checking it, it should be safe, it is just a matter of efficiency

kishore-nori replied:
> Do these ratios keep asympotatically fixed as you refine the mesh?
I ll check and update.

kishore-nori replied:
On refinement the the new version runs in the same time but the earlier one becomes slower and more memory intensive.

amartinhuertas commented on src/CellData/DiracDeltas.jl:
```diff
@@ -65,14 +70,63 @@ function DiracDelta{0}(model::DiscreteModel;tags)
+  cell = _point_to_cell!(cache, p)
+  trianv = TriangulationView(trian,[cell])
+  pquad = GenericQuadrature([p],[one(T)])
+  pmeas = Measure(CellQuadrature([pquad],[[p]],[[one(T)]],trianv,PhysicalDomain(),PhysicalDomain()))
```
to be on the safe side w.r.t. memory allocations, better not to create `[p]` and `[one(T)]` twice.

why do we have to pass `[[p]]` to `CellQuadrature` constructor? Couldn't it get this info from `pquad` ? (perhaps a missing constructor in Gridap?)

kishore-nori replied:
Yes true, I was thinking the same, may be a missing constructor, specifically for `data_domain_style` being `PhysicalDomain()`, the problem is we can't use the `compress_cell_data` for `AbstractArrays` only for `Fill` and `CompressedArray`, which is the usual quadrature case. Yes I ll remove the repetition and reuse.

amartinhuertas replied:
> Yes true, I was thinking the same, may be a missing constructor, specifically for data_domain_style being PhysicalDomain(), the problem is we can't use the compress_cell_data for AbstractArrays only for Fill and CompressedArray, which is the usual quadrature case. Yes I ll remove the repetition and reuse.

Ok, I see. The constructor is missing, but if we added it, we would not be able to leverage it in our particular context (we do not have Fill or CompressedArrays to represent the quadrature points and weights).

amartinhuertas reviewed (changes requested):
Hi @kishore-nori ! nice work! I finished with my review. Please answer/rebut my comments.

amartinhuertas commented on src/CellData/DiracDeltas.jl:
```diff
@@ -65,14 +70,65 @@ function DiracDelta{0}(model::DiscreteModel;tags)
+  trian = Triangulation(model)
+  cache = _point_to_cell_cache(KDTreeSearch(),trian)
+  cell = _point_to_cell!(cache, p)
+  trianv = TriangulationView(trian,[cell])
```
use `view` Julia method instead of `TriangulationView`. 

This way we do not need to export `TriangulationView`.

amartinhuertas commented on src/Geometry/Geometry.jl:
```diff
@@ -102,6 +102,7 @@ export compress_ids
 export UnstructuredGridTopology
 
 export Triangulation
+export TriangulationView
```
see https://github.com/gridap/Gridap.jl/pull/837/files#r996625877

kishore-nori replied:
done!

amartinhuertas commented on src/CellData/DiracDeltas.jl:
```diff
@@ -65,14 +70,65 @@ function DiracDelta{0}(model::DiscreteModel;tags)
+  points = map(i->pvec[cell_points[i]], 1:length(cell_ids))
+  weights_x_cell = collect.(Fill.(one(T),length.(cell_points)))
+  pquad = map(i -> GenericQuadrature(points[i],weights_x_cell[i]), 1:length(cell_ids))
+  trianv = TriangulationView(trian,cell_ids)
```
https://github.com/gridap/Gridap.jl/pull/837/files#r996625877

kishore-nori replied:
done!

amartinhuertas commented on src/CellData/DiracDeltas.jl:
```diff
@@ -65,14 +70,65 @@ function DiracDelta{0}(model::DiscreteModel;tags)
+  GenericDiracDelta{0,D,NotGridEntity}(trianv,pmeas)
+end
+
+function evaluate!(cache,d::GenericDiracDelta{0,Dt,NotGridEntity},f::Function) where Dt
```
Remove this method.

kishore-nori replied:
done :)
